src/main/scala/is/hail/expr/FunctionRegistry.scala

```diff
@@ -2785,6 +2957,208 @@ object FunctionRegistry {
+      i += 1
+    }
+    a: IndexedSeq[java.lang.Integer]
+  }, null)(arrayHr(boxedInt32Hr), int32Hr, arrayHr(boxedInt32Hr))
+
+  register("%", { (x: Int, ys: IndexedSeq[java.lang.Integer]) =>
+    val a = new Array[java.lang.Integer](ys.length)
+    var i = 0
+    while (i < a.length) {
+      val y = ys(i)
+      if (y != null)
+        a(i) = java.lang.Math.floorMod(x, y)
+      i += 1
+    }
+    a: IndexedSeq[java.lang.Integer]
+  }, null)(int32Hr, arrayHr(boxedInt32Hr), arrayHr(boxedInt32Hr))
+
+  register("%", { (xs: IndexedSeq[java.lang.Integer], ys: IndexedSeq[java.lang.Integer]) =>
+    if (xs.length != ys.length)
+      fatal(
+        s"""Cannot apply operation '%' to arrays of unequal length:
+           |  Left: ${ xs.length } elements, [${ xs.mkString(", ") }]
+           |  Right: ${ ys.length } elements [${ ys.mkString(", ") }]""".stripMargin)
+
+    val a = new Array[java.lang.Integer](xs.length)
+    var i = 0
+    while (i < a.length) {
+      val x = xs(i)
+      val y = ys(i)
+      if (x != null && y != null)
+        a(i) = java.lang.Math.floorMod(xs(i), ys(i))
+      i += 1
+    }
+    a: IndexedSeq[java.lang.Integer]
+  }, null)(arrayHr(boxedInt32Hr), arrayHr(boxedInt32Hr), arrayHr(boxedInt32Hr))
+
+  register("%", { (xs: IndexedSeq[java.lang.Long], y: Long) =>
+    val a = new Array[java.lang.Long](xs.length)
+    var i = 0
+    while (i < a.length) {
+      val x = xs(i)
+      if (x != null)
+        a(i) = java.lang.Math.floorMod(x, y)
+      i += 1
+    }
+    a: IndexedSeq[java.lang.Long]
+  }, null)(arrayHr(boxedInt64Hr), int64Hr, arrayHr(boxedInt64Hr))
+
+  register("%", { (x: Long, ys: IndexedSeq[java.lang.Long]) =>
+    val a = new Array[java.lang.Long](ys.length)
+    var i = 0
+    while (i < a.length) {
+      val y = ys(i)
+      if (y != null)
+        a(i) = java.lang.Math.floorMod(x, y)
+      i += 1
+    }
+    a: IndexedSeq[java.lang.Long]
+  }, null)(int64Hr, arrayHr(boxedInt64Hr), arrayHr(boxedInt64Hr))
+
+  register("%", { (xs: IndexedSeq[java.lang.Long], ys: IndexedSeq[java.lang.Long]) =>
+    if (xs.length != ys.length)
+      fatal(
+        s"""Cannot apply operation '%' to arrays of unequal length:
+           |  Left: ${ xs.length } elements, [${ xs.mkString(", ") }]
+           |  Right: ${ ys.length } elements [${ ys.mkString(", ") }]""".stripMargin)
+
+    val a = new Array[java.lang.Long](xs.length)
+    var i = 0
+    while (i < a.length) {
+      val x = xs(i)
+      val y = ys(i)
+      if (x != null && y != null)
+        a(i) = java.lang.Math.floorMod(xs(i), ys(i))
+      i += 1
+    }
+    a: IndexedSeq[java.lang.Long]
+  }, null)(arrayHr(boxedInt64Hr), arrayHr(boxedInt64Hr), arrayHr(boxedInt64Hr))
+
+  register("%", { (xs: IndexedSeq[java.lang.Float], y: Float) =>
+    val a = new Array[java.lang.Float](xs.length)
+    var i = 0
+    while (i < a.length) {
+      val x = xs(i)
+      if (x != null) {
+        val t = x % y
+        if (x >= 0 && y > 0 || x <= 0 && y < 0 || t == 0)
```

Comment:
you could pull `y > 0` out of the loop as `val yPos = y > 0` but probably doesn't make much difference.

Reply:
yeah, doesn't look like it'll make much difference. not worth the effort to go back in.

Reply:
I suspect the JVM JITs this sort of stuff.